### PR TITLE
Adding restart backoff and return in case of an error

### PR DESCRIPTION
**Pull request by amitschendel**

### CODE REVIEW

File: pkg/watcher/watcher.go [hunk 1]
```diff
@@ -104,9 +105,19 @@ func (w *Watcher) Start(notifyF WatchNotifyFunctions, gvr schema.GroupVersionRes
...
+							break
+						}
+						log.Printf("watcher restart error: %v, on object: %+v, retrying...", err, gvr)
+						time.Sleep(time.Second * time.Duration(i*2)) // Exponential backoff
+					}
+
 					if err != nil {
-						log.Printf("watcher restart error: %v, on object: %+v", err, gvr)
+						log.Printf("watcher restart failed after 5 attempts, on object: %+v", gvr)
+						return
```
Comment (Bezbran) [inline pkg/watcher/watcher.go:120]: What happens next?
maybe we better to restart the process

**Review by Bezbran — CHANGES_REQUESTED** [review]

File: pkg/watcher/watcher.go (hunk 1 above)
Comment (amitschendel) [reply pkg/watcher/watcher.go]: done

File: pkg/watcher/watcher.go [hunk 2]
```diff
@@ -104,9 +105,19 @@ func (w *Watcher) Start(notifyF WatchNotifyFunctions, gvr schema.GroupVersionRes
...
+							break
+						}
+						log.Printf("watcher restart error: %v, on object: %+v, retrying...", err, gvr)
+						time.Sleep(time.Second * time.Duration(i*2)) // Exponential backoff
+					}
+
 					if err != nil {
-						log.Printf("watcher restart error: %v, on object: %+v", err, gvr)
+						// If the watcher restart failed after 5 attempts, exit so the pod can be restarted.
+						log.Fatalf("watcher restart failed after 5 attempts: %v", err)
```
Comment (Bezbran) [inline pkg/watcher/watcher.go:120]: From a software design perspective, this decision is not within the scope of the package's purpose. It should be up to the caller of the function to decide what action to take.

**Review by Bezbran — CHANGES_REQUESTED** [review]

File: pkg/watcher/watcher.go (hunk 2 above)
Comment (amitschendel) [reply pkg/watcher/watcher.go:120]: What other options do we have?

Comment (amitschendel) [reply pkg/watcher/watcher.go]: done

**Review by Bezbran — APPROVED** [review]

**Pull request merged**